add emapplot.compareClusterResult function

huerqiang opened the pull request.

GuangchuangYu commented:
```
y <- as.data.frame(x)
  if (is.numeric(n)) {
        y <- y[1:n,]
    } else {
        y <- y[match(n, y$Description),]
        n <- length(n)
    }

```

你看看<https://mp.weixin.qq.com/s/VW4LvtQqXumrs6SMvwGgvQ>，`if`这一段没有考虑到这个需求，而`else`里面倒是有。

GuangchuangYu commented:
`geom_scatterpie_legend2`和`scatterpie::geom_scatterpie_legend`有什么差别？

GuangchuangYu commented:
```
merge_compareClusterResult <- function(yy) {
	yy_union<- yy[!duplicated(yy$ID),]
	rownames(yy_union) <- yy_union$ID
	# Make a for loop to merge genes
	for(i in seq_len(dim(yy)[1])) {
	    gene1 <- unlist(strsplit(yy[i,"geneID"],"/", fixed = TRUE))
		gene2 <- unlist(strsplit(yy_union[yy[i,"ID"],"geneID"],"/", fixed = TRUE))
		genes <- union(gene1,gene2)
		gene_ID_uni <- paste(genes,collapse="/")
		yy_union[yy[i,"ID"],"geneID"] <- gene_ID_uni	
	}
	return(yy_union)
}

```

这里面`yy`和`yy_union`怎么可能用`i`来取子集还能对得上？唯一的解释就是我第一条评论中，你跑起来是`if`语句里的`y`。

GuangchuangYu commented:
构建图那块代码，最好分离出来，和`emapplot.enrichResult`都使用同样的函数，第一不要copy代码，第二构建图这个方法可能会变，到时候只需要改动相应函数即可，而不是散在的多个拷贝。

GuangchuangYu commented:
构建图那块代码，最好分离出来，和`emapplot.enrichResult`都使用同样的函数，第一不要copy代码，第二构建图这个方法可能会变，到时候只需要改动相应函数即可，而不是散在的多个拷贝。

GuangchuangYu commented:
再者，代码要注释，特别是画饼图这块。

huerqiang commented:
> `geom_scatterpie_legend2`和`scatterpie::geom_scatterpie_legend`有什么差别？

geom_scatterpie_legend2和scatterpie::geom_scatterpie_legend相比只改了一句代码：dd$label <- (dd$r)^2*500。 因为我想让标签表示实际的基因数量，而直接用scatterpie::geom_scatterpie_legend的话它标注的是饼图的半径，而饼图半径是ID_Cluster_mat$radius <- sqrt(aa$size / 500) 转化而来的

huerqiang commented:
> ```
> merge_compareClusterResult <- function(yy) {
> 	yy_union<- yy[!duplicated(yy$ID),]
> 	rownames(yy_union) <- yy_union$ID
> 	# Make a for loop to merge genes
> 	for(i in seq_len(dim(yy)[1])) {
> 	    gene1 <- unlist(strsplit(yy[i,"geneID"],"/", fixed = TRUE))
> 		gene2 <- unlist(strsplit(yy_union[yy[i,"ID"],"geneID"],"/", fixed = TRUE))
> 		genes <- union(gene1,gene2)
> 		gene_ID_uni <- paste(genes,collapse="/")
> 		yy_union[yy[i,"ID"],"geneID"] <- gene_ID_uni	
> 	}
> 	return(yy_union)
> }
> ```
> 
> 这看着就觉得很怪，你解释一下吧。

因为compareClusterResult的结果中，同个ID在多个Cluster中都有一个geneID，所有我想用这个ID在所有Cluster中的geneID的并集作为富集在这个ID的基因集合

huerqiang commented:
关于“这里面yy和yy_union怎么可能用i来取子集还对得上”的回复：这里循环中的i是对yy的每一行进行遍历，然后将它的geneID合并到yy_union对应ID的geneID中。

GuangchuangYu commented:
逐个来解决吧。
1. `geom_scatterpie_legend` 去`scatterpie`包搞个PR，用半径还是size，通过参数控制。
    - <https://github.com/YuLab-SMU/enrichplot/pull/26#issuecomment-546565263>
2. 我提到的公众号那篇文章好好读读，把这个因素考虑进去。
    - <https://github.com/YuLab-SMU/enrichplot/pull/26#issuecomment-546388059>
3. `merge_compareClusterResult`函数的问题，讲清楚rational在哪里，为什么要做这个事情。再者你这个实现效率过慢，不断在切字符串和贴字符串，只需要切一次，贴一次。

huerqiang commented:
回复1：我把geom_scatterpie_legend 添加参数控制后，不清楚为什么起不了作用，我在用enrichplot::emapplot()画饼图的时候一直显示的是实际大小，而不是基因数量，想请Y叔指导一下。
不知道是不是enrichplot::emapplot.compareClusterResult泛型函数的参数我使用的不正确。另外
我在用getAnywhere("emapplot.compareClusterResult")观察我的函数时，最下面显示：
Warning message:
In findGeneric(f, envir) : 'emapplot'是正式同属函数; 很有可能找不到S3方法

回复2：我把clusterProfiler:::fortify对compareClusterResult结果的预处理用在了emapplot.compareClusterResult中了。
回复3：merge_compareClusterResult函数的rational在于：饼图的面积我觉得使用富集到这个GO ID的所有cluster的基因的并集表示，因此进行合并。 已经对这个函数进行优化，先全部切开，然后再一起合并。
回复4：做函数的时候出现了一点问题，还需要点时间进行处理.
回复5：添加了注释

GuangchuangYu commented:
问题一个个来。

1. 像这种对`scatterpie`修改的问题，你应该问自己改完之后，它自身行不行，而不是改完之后，你基于此的函数行不行，前面是一步，后面是两步。

>geom_scatterpie_legend2和scatterpie::geom_scatterpie_legend相比只改了一句代码：dd$label <- (dd$r)^2*500。 因为我想让标签表示实际的基因数量，而直接用scatterpie::geom_scatterpie_legend的话它标注的是饼图的半径，而饼图半径是ID_Cluster_mat$radius <- sqrt(aa$size / 500) 转化而来的

+ 我完全没看出来500是在哪里操作了。
+ 你要的效果，完全不需要改动代码，`scatterpie`本身就是支持的。但凡你有看过文档<https://cran.r-project.org/web/packages/scatterpie/vignettes/scatterpie.html>.

huerqiang commented:
> 问题一个个来。
> 
> 1. 像这种对`scatterpie`修改的问题，你应该问自己改完之后，它自身行不行，而不是改完之后，你基于此的函数行不行，前面是一步，后面是两步。
> 
> > geom_scatterpie_legend2和scatterpie::geom_scatterpie_legend相比只改了一句代码：dd$label <- (dd$r)^2*500。 因为我想让标签表示实际的基因数量，而直接用scatterpie::geom_scatterpie_legend的话它标注的是饼图的半径，而饼图半径是ID_Cluster_mat$radius <- sqrt(aa$size / 500) 转化而来的
> 
> * 我完全没看出来500是在哪里操作了。
> * 你要的效果，完全不需要改动代码，`scatterpie`本身就是支持的。但凡你有看过文档https://cran.r-project.org/web/packages/scatterpie/vignettes/scatterpie.html.

我按照您的文档的信息后，使用labeller=function(x)...  参数，画出了正常的图。接下来我继续做“构建图那块代码，最好分离出来，和emapplot.enrichResult都使用同样的函数，第一不要copy代码，第二构建图这个方法可能会变，到时候只需要改动相应函数即可”的工作。

huerqiang commented:
回复4：构建图函数
我新建了一个函数graph_build()来建graph网络图中的edge，emapplot.enrichResult()和emapplot.compareClusterResult()都对它进行了调用

GuangchuangYu commented:
`
clusterr <- sapply(X = y$Cluster, FUN = function(dd){unlist(strsplit(as.character(dd),"\n"))[1]})
y$Cluster <- clusterr
```

第一不要在生产代码中使用`sapply`，它有可能返回list（用lapply），有可能返回vector（用vapply）。

上面的代码干的事情其实是：

```
y$Cluster = sub("\n.*", "", y$Cluster)
```

我这一行，不单简单容易看，而且你如果去测试速度的话，是你那个1/30的运行时间。

GuangchuangYu commented:
```
merge_compareClusterResult <- function(yy) {
	yy_union<- yy[!duplicated(yy$ID),]
	rownames(yy_union) <- yy_union$ID
	# Make a for loop to merge genes
	for(i in seq_len(dim(yy)[1])) {
	    gene1 <- unlist(strsplit(yy[i,"geneID"],"/", fixed = TRUE))
		gene2 <- unlist(strsplit(yy_union[yy[i,"ID"],"geneID"],"/", fixed = TRUE))
		genes <- union(gene1,gene2)
		gene_ID_uni <- paste(genes,collapse="/")
		yy_union[yy[i,"ID"],"geneID"] <- gene_ID_uni	
	}
	return(yy_union)
}


merge_compareClusterResult <- function(yy) {
    #one strsplit, one paste

	

	list_yy <- lapply(X = yy[,"geneID"],FUN=function(aa) {unlist(strsplit(as.character(aa),"/", fixed = TRUE))})
	yy_union<- yy[!duplicated(yy$ID),]
	list_yy_union <- lapply(X = yy_union[,"geneID"],FUN=function(aa) {unlist(strsplit(as.character(aa),"/", fixed = TRUE))})
	names(list_yy_union) <- yy_union$ID
	rownames(yy_union) <- yy_union$ID
	
	# Make a for loop to merge genes
	for(i in seq_len(dim(yy)[1])) {
	    #gene1 <- unlist(strsplit(yy[i,"geneID"],"/", fixed = TRUE))
		gene1 <- list_yy[[i]]
		#gene2 <- unlist(strsplit(yy_union[yy[i,"ID"],"geneID"],"/", fixed = TRUE))
		gene2 <- list_yy_union[yy[i,"ID"]][[1]]
		genes <- union(gene1,gene2)
		gene_ID_uni <- paste(genes,collapse="/")
		yy_union[yy[i,"ID"],"geneID"] <- gene_ID_uni	
	}
	return(yy_union)
}
```

代码出现两个版本，太不走心。

第二个版本，好意思说one strsplit, one paste？

你的`paste`放在`for`里面，怎么可能one paste?

而且这段代码根本就是错误的，每一次paste出来都是新的，然后覆盖原来的，而不是整合在一起。

GuangchuangYu commented:
x_loc1 <- min(ID_Cluster_mat$x)
y_loc1 <- min(ID_Cluster_mat$y)
x_loc2 <- min(ID_Cluster_mat$x)
y_loc2 <- min(ID_Cluster_mat$y)+0.1*(max(ID_Cluster_mat$y)-min(ID_Cluster_mat$y))

x_loc1和x_loc2是一样的。y_loc2前面的和减后面的也是一样的。

GuangchuangYu commented:
geneSets <- geneInCategory(x) 

这一句调用的太快，显然后面建图的时候，基于这个geneSets来计数，数目是显著少太多。

GuangchuangYu commented:
enrichplot不能import clusterProfiler，因为clusterProfiler需要enrichplot的画图功能，这样会出现依赖包闭环。需要把clusterProfiler中关于compareClusterResult的画图部分移过来，包括fortify, dotplot等。

GuangchuangYu commented:
我重新写了`merge_compareClusterResult`，改掉一些bugs。你好好看一下。<https://github.com/YuLab-SMU/enrichplot/commit/f0e767d6242ca0fa6cc9f465a9a5bd2ab54612f3>.


每次只解决一个点，自己仔细看清楚了再PR，当然即便你PR了，也要仔细再检查，再反复看看有什么问题以及改进的地方。

你自己觉得完全没问题了，再跟我说，我再来看。

因为从这几次的PR来看，有很多，多看两眼就可以发现的问题。

1. `graph_build`我重命名为`emap_graph_build`，你还需要再改改，构图的输出应该是`igraph`对象，网络图对象和画图要分开。
2. 另外最终的图也要再修一些细节，比如现在node label是没有的。
3. 以及上面提到的把clusterProfiler的代码移过来。


所以又有三个事情要做。

GuangchuangYu commented:
第三点，我已经解决了。因为你的commit，**让包不可用！！！**所以得赶紧解决，起码包得可以用。

这也给你提个醒，每次改完代码，要跑一下`devtools::check()`检查一下。

像这个依赖包闭环，在`enrichplot`中用自己的函数使用了`enrichplot:::`这种骚操作，都会使包无法编译，无法安装，无法使用。
